### ISSUE TITLE
Question: Usage model with multiple inputs

### ISSUE THREAD

**Issue by lebionick** [KevinMusgrave/pytorch-metric-learning#37]
Hello,
I want to train pre-trained BERT model, which accepts three arguments (token_ids, token_types, attention_mask), all of them are tensors (N x L). As far as I understand from your source code, during training with `trainer` instance, I am able to put only one tensor into a model. 
What would you recommend for me: 1. wrap model into another model, that accepts stacked tensors and splits them 2. edit source code 3. something else? Or may be I am wrong, please correct me, if it so

**Comment by KevinMusgrave** [comment]
@lebionick  I think there's two options:

1. Wrap the model. This is probably the easiest and quickest solution at the moment.

2. Extend the trainer you're using, and override the [get_trunk_output](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/master/src/pytorch_metric_learning/trainers/base_trainer.py#L137) method. Maybe something like this:

```python
from pytorch_metric_learning import trainers

class YourTrainer(trainers.MetricLossOnly):
    def get_trunk_output(self, data):
        return self.models["trunk"](*data)
```

Do token_ids, token_types, and attention_mask come from a dataset? If so, you'll probably need to pass in ```data_and_label_getter``` when you initialize the trainer.

**Comment by lebionick** [comment]
@KevinMusgrave Thank you for your suggestions! I will try it and write here if it worked or not.

**Comment by lebionick** [comment]
Yes, it worked nice, also I overrode `tester` analogously: 
``` python
class CustomGlobalEmbeddingSpaceTester(testers.GlobalEmbeddingSpaceTester):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

    def get_embeddings_for_eval(self, trunk_model, embedder_model, input_imgs):
        input_imgs = dict_to_cuda(input_imgs)
        trunk_output = trunk_model.forward(**input_imgs)
        return embedder_model(trunk_output)
```
trunk_model here is `nemo.collection.nlp.nm.trainables.huggingface.BERT` and only uses method `forward` for inference; my dataset returns `Tuple[Dict[str, torch.Tensor], int]`, so I write my own collate function:
``` python
def custom_collate_fn(batch):
    # list of tuples of dict and int
    data_keys = batch[0][0].keys()
    new_data = {}
    for key in data_keys:
        key_tensors = [x[0][key] for x in batch]
        batch_max_length = get_max_len(key_tensors)
        key_tensors = [add_padding(t, batch_max_length) for t in key_tensors]
        batch_tensor = torch.cat(key_tensors, 0)
        new_data[key] = batch_tensor
    labels = [x[1] for x in batch]
    new_labels = torch.LongTensor(labels)
    return new_data, new_labels
```
(I add here padding, because input sequences (text) are different in size)  

P.S. I have put here the code just in case someone else will need it :)

**Comment by KevinMusgrave** [comment]
@lebionick Nice! By the way, I'm not familiar with these NLP models, and there's another issue [regarding the output size of BERT](https://github.com/KevinMusgrave/pytorch-metric-learning/issues/29#issue-588174987). Basically for a batch size of N, he is getting an output size (N, D1, D2), rather than (N, D). I'm not sure how to deal with that during training or evaluation. If you happen to have any ideas about this, let me know (or post in that issue).
